python3Packages.pytorch: 1.2.0 -> 1.4.1, python3Packages.ignite: 0.2.1 -> 0.3.0

stites opened the pull request:
###### Motivation for this change
Update pytorch to 1.3.1. A detailed commit history [can be seen here](https://github.com/stites/pytorch-world/pull/14).
Check phase verified on python36.pytorch, python36.pytorchWithMkl, python36.pytorchWithCuda10, python36.pytorchWithCuda10Mkl.

Cachix pending (my machine with the keys is down).

Relevant changelog:
- `buildDocs` flag added
- `buildNamedTensor` is now true by default
- adds experimental `useNixProtobuf` but disables this functionality.
- uses CMake directly for pytorch builds

###### To build

I believe the following should work:
```
git remote add --fetch stites git@github.com:stites/nixpkgs.git
git checkout --track stites pytorch-1.3.1
nix-shell -I nixpkgs=$(pwd) -p "python37.withPackages(ps: with ps; [ pytorch ])"
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @teh @thoughtpolice @stites @tscholak @bhipple

stites committed:
pytorch: 1.2.0 -> 1.4.1

Co-authored-by: Benjamin Hipple <bhipple@protonmail.com>

stites commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -165,38 +202,50 @@ in buildPythonPackage rec {
     click
     numpy
     pyyaml
+    psutil
```
Looks like this is only a test-time dependency.

jonringer commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -1,30 +1,33 @@
                         in majorIs == "9" || majorIs == "10");
 
+# Using nix's protobuf instead of the bundled version with pytorch is considered broken.
+assert useNixProtobuf == false;
```
then why allow the option if it's always wrong?

stites replied:
Yeah... this was just hopeful code. I should have cleaned this up. Updating now.

jonringer commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -165,38 +202,49 @@ in buildPythonPackage rec {
+      # ^^^^^^^^^^^^ NOTE: while test_dataloader does return errors, these are acceptable errors and do not interfere with the build
+
+      # tensorboard has acceptable failures for pytorch 1.3.x due to dependencies on tensorboard-plugins
+      (with lib.versions; optionalString (major version == "1" && minor version == "3") "tensorboard")
```
```suggestion
      (with lib.versions; optionalString (majorMinor version == "1.3" ) "tensorboard")
```

stites replied:
ah! I just added this refactor (as well as to the mkl version above). Not yet used to adding code via the github UI.

stites reviewed (commented):
Removed the nix-protobuf code (it will just live in pytorch-world for now). Rebuilding everything again -- I just had to deal with some faulty RAM slots so updates will have a slightly faster turnaround again.

veprbl commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -122,18 +120,28 @@ in buildPythonPackage rec {
+  BUILD_NAMEDTENSOR = true;
+  BUILD_DOCS = buildDocs;
+  USE_MKLDNN = mklSupport;
+  preBuild = with lib.strings; concatStringsSep "\n" [
```
```suggestion
  preBuild = ''
```

veprbl commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -122,18 +120,28 @@ in buildPythonPackage rec {
+  BUILD_DOCS = buildDocs;
+  USE_MKLDNN = mklSupport;
+  preBuild = with lib.strings; concatStringsSep "\n" [
+    ''export CMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH:${if mklSupport then mkl else openblas}"''
```
Shouldn't cmake's setupHook do this allready? Just add `mkl` or `openblas` to the `buildInputs`.

stites replied:
Unfortunately, as you noted in https://github.com/NixOS/nixpkgs/pull/75827#discussion_r360659577, pytorch has it's own custom cmake configuration tooling.

From what I observed, you couldn't just add `mkl` or `openblas` to the buildinputs for pytorch. It will build, but the dependencies won't be hooked in and you will only observe failures during test time.

veprbl replied:
I am still not convinced. We do set the CMAKE_PREFIX_PATH:
https://github.com/NixOS/nixpkgs/blob/bc555a24afbe0be95a90aee6cfccf626f4d6ac08/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L123
https://github.com/NixOS/nixpkgs/blob/bc555a24afbe0be95a90aee6cfccf626f4d6ac08/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L1-L3
You might need to move `(if mklSupport then mkl else openblas)` from `nativeBuildInputs` to `buildInputs` to make it work.

stites replied:
You're absolutely correct and this did work out -- I guess I was expecting nativeBuildInputs to be added to the cmake path

veprbl commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -122,18 +120,28 @@ in buildPythonPackage rec {
+  preBuild = with lib.strings; concatStringsSep "\n" [
+    ''export CMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH:${if mklSupport then mkl else openblas}"''
+    "python setup.py build --cmake-only"
+    "${cmake}/bin/cmake build -DBLAS=${if mklSupport then "MKL" else "OpenBLAS"}"
```
"BLAS" should be settable from the environment?
https://github.com/pytorch/pytorch/blob/700109eb630b79fd65cb93becb7f2d14f93bdb5c/tools/setup_helpers/cmake.py#L228

stites replied:
Looks like this was captured via `buildInputs` as you mentioned!

FRidh commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -122,18 +120,28 @@ in buildPythonPackage rec {
+  USE_MKLDNN = mklSupport;
+  preBuild = with lib.strings; concatStringsSep "\n" [
+    ''export CMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH:${if mklSupport then mkl else openblas}"''
+    "python setup.py build --cmake-only"
```
```${python.interpreter}```

Are wheels not yet supported?

stites replied:
Wheels are supported! That's how pytorch is distributed with pip and various configurations can be found at [pytorch.org](https://pytorch.org).  I just always assumed that building from source was the preferred nix way.

Is there a good example which depends on third-party C/C++-libraries you can point me to? searching the github repository seems noisy.

The python section makes it look like I should pass `format="wheel"` -- is it possible to simply do this and fetch the source with `fetchPypi`?

jonringer replied:
In general yes, it's preferred to build from source. On reason to keep the source build is to be able to configure the package for things like CUDA or changing blas implementation. Also, tests are usually distributed with the source distribution.

But if you can correctly link a wheel, and it works as intended then that's definitely a possible route.

FRidh replied:
I meant *building* a wheel.

FRidh replied:
but I see now it's just a preBuild step

jonringer replied:
oh, you mean `python setup.py build bdist_wheel`, xD

FRidh commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -165,38 +173,49 @@ in buildPythonPackage rec {
+    # MKL 2019.5-only workaround. See: https://github.com/NixOS/nixpkgs/issues/75611
+    (optionalString (mklSupport && majorMinor mkl.version == "2019.5") "KMP_INIT_AT_FORK=FALSE ")
+    cudaStubEnv
+    "python test/run_test.py"
```
${python.interpreter}

danieldk commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -122,18 +120,28 @@ in buildPythonPackage rec {
+
+  BUILD_NAMEDTENSOR = true;
+  BUILD_DOCS = buildDocs;
+  USE_MKLDNN = mklSupport;
```
It seems that `mkl`, `mklSupport` and MKLDNN are entangled in this derivation. However, it is possible to build MKLDNN without MKL. MKDNN *can* use an external BLAS implementation, which can be MKL.

So, it seems that there should be a separate `mklDnnSupport` flag, which should probably be set to `true` by default?

stites replied:
I'll have to double-check this -- in a former derivation, I recall them being tightly coupled in the pytorch build.

danieldk replied:
I compiled this PR with the following change:

```
diff --git a/pkgs/development/python-modules/pytorch/default.nix b/pkgs/development/python-modules/pytorch/default.nix
index bb98f4ecffc..817141abc10 100644
--- a/pkgs/development/python-modules/pytorch/default.nix
+++ b/pkgs/development/python-modules/pytorch/default.nix
@@ -1,6 +1,7 @@
 { stdenv, fetchurl, fetchgit, buildPythonPackage, python, pythonOlder,
   cudaSupport ? false, cudatoolkit ? null, cudnn ? null, nccl ? null, magma ? null,
   mklSupport ? false, mkl ? null, openblas ? null,
+  mklDnnSupport ? true,
   openMPISupport ? false, openmpi ? null,
   buildBinaries ? false,
   buildDocs ? false,
@@ -135,7 +136,7 @@ in buildPythonPackage rec {
 
   BUILD_NAMEDTENSOR = true;
   BUILD_DOCS = buildDocs;
-  USE_MKLDNN = mklSupport;
+  USE_MKLDNN = mklDnnSupport;
   preBuild = with lib.strings; concatStringsSep "\n" [
     ''export CMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH:${if mklSupport then mkl else openblas}"''
     "python setup.py build --cmake-only"
```

And it works as expected, without any dependency on Intel MKL.

stites replied:
Thanks for the nudge, @danieldk -- I've amended this commit with your suggestion.

bcdarwin commented:
fwiw, 1.4 is now out.

danieldk commented:
> fwiw, 1.4 is now out.

There may soon also be a 1.4.1 release that is compatible with gcc 9:

https://github.com/pytorch/pytorch/issues/32277

stites commented:
@bcdarwin @danieldk -- I've bumped this PR to 1.4.0. I'm fairly confident this derivation remains bug-free since I ran the full test suite on the alpha branch a few days before the 1.4.0 release. Unfortunately, I don't have time to run the entire build matrix again.

jonringer commented:
@GrahamcOfBorg build python2Packages.pytorch
@GrahamcOfBorg build python3Packages.pytorch
@GrahamcOfBorg build python38Packages.pytorch

jonringer commented:
please address failures :(

jonringer commented:
looks like a lot of failures are related to `-fpermissive`. I forgot the exact steps, but you can lower it from an error to a warning

bcdarwin commented:
When building on x86 Ubuntu with Cuda support, the package builds but is broken:

```
$ nix-shell -I nixpkgs=~/nixpkgs -p 'python3.withPackages (ps: with ps; [ pytorch ])' --pure
...
$ python3 -c 'import torch'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/nix/store/3mkr7x854wslwpdl0zv9kg4k1k8icirc-python3-3.7.5-env/lib/python3.7/site-packages/torch/__init__.py", line 81, in <module>
    from torch._C import *
ImportError: /nix/store/3mkr7x854wslwpdl0zv9kg4k1k8icirc-python3-3.7.5-env/lib/python3.7/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch4cuda4nccl6detail16throw_nccl_errorE12ncclResult_t
```

Possibly setting `USE_SYSTEM_NCCL=true` doesn't work with the current NCCL/Pytorch.

Seems fine without Cuda but didn't run the test suite.

bcdarwin commented:
Also, a bunch of test failures in `test_tensorboard.py`:
```
======================================================================
FAIL: test_audio (__main__.TestTensorBoardSummary)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_tensorboard.py", line 393, in test_audio
    self.assertTrue(compare_proto(summary.audio('dummy', tensor_N(shape=(42,))), self))
AssertionError: False is not true

======================================================================
FAIL: test_histogram_auto (__main__.TestTensorBoardSummary)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_tensorboard.py", line 399, in test_histogram_auto
    self.assertTrue(compare_proto(summary.histogram('dummy', tensor_N(shape=(1024,)), bins='auto', max_bins=5), self))
AssertionError: False is not true

======================================================================
FAIL: test_histogram_doane (__main__.TestTensorBoardSummary)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_tensorboard.py", line 405, in test_histogram_doane
    self.assertTrue(compare_proto(summary.histogram('dummy', tensor_N(shape=(1024,)), bins='doane', max_bins=5), self))
AssertionError: False is not true

======================================================================
FAIL: test_histogram_fd (__main__.TestTensorBoardSummary)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_tensorboard.py", line 402, in test_histogram_fd
    self.assertTrue(compare_proto(summary.histogram('dummy', tensor_N(shape=(1024,)), bins='fd', max_bins=5), self))
AssertionError: False is not true

======================================================================
FAIL: test_mesh (__main__.TestTensorBoardSummary)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_tensorboard.py", line 448, in test_mesh
    self.assertTrue(compare_proto(mesh, self))
AssertionError: False is not true

----------------------------------------------------------------------
Ran 45 tests in 0.534s

FAILED (failures=5, skipped=3)
Traceback (most recent call last):
  File "test/run_test.py", line 455, in <module>
    main()
  File "test/run_test.py", line 448, in main
    raise RuntimeError(message)
RuntimeError: test_tensorboard failed!
builder for '/nix/store/s6fi2ryiy6hrzix7m9gvglvkz0y5dai4-python3.7-pytorch-1.4.0.drv' failed with exit code 1
```

danieldk commented:
> looks like a lot of failures are related to `-fpermissive`. I forgot the exact steps, but you can lower it from an error to a warning.

These come from incompatibility with gcc 9 (see the PyTorch bug mentioned above), until they release a new version with the fix, PyTorch can only be compiled with gcc < 9. It should work fine in the CUDA case due to `CC` being set for CUDA.

stites commented:
@bcdarwin I was sort of hoping the tensorboard errors would be solved by pytorch 1.4.0. These don't seem like plugin errors (1.3 had some sort of mesh plugin which was breaking tests, documented here https://github.com/NixOS/nixpkgs/pull/75827/files#diff-84b29016f8c93846ea69f3885d709770R215 ). it's been a while since I looked at the broken tensorboard tests -- I did try to bump protobuf and tensorboard in different PRs (which I should probably check in on) but I recall finding out that pytorch actually just uses tensorboard 1.x and I rolled back the nix-protobuf dependency in this PR, so that seems like the next place to investigate. I will be unable to do this.

I haven't run into nccl problems on NixOS -- as you pointed out, likely `USE_SYSTEM_NCCL=true` is breaking the Ubuntu build. Here, I think hoisting a `useSystemNccl ? true` into the build arguments might suffice -- I can't test this functionality, but I'm pushing up an amended commit now.

@danieldk  @jonringer  it looks like `-fpermissive` warnings are not suppressible -- is there an implicit `-Werr` in nix, or do I need to wait until pytorch-1.4.1?

danieldk commented:
> @danieldk @jonringer it looks like `-fpermissive` warnings are not suppressible -- is there an implicit `-Werr` in nix, or do I need to wait until pytorch-1.4.1?

In my local PyTorch derivation I use something like:

```
  preConfigure = (if cudaSupport then ''
    export TORCH_CUDA_ARCH_LIST="${lib.strings.concatStringsSep ";" final_cudaArchList}"
    export CC=${cudatoolkit.cc}/bin/gcc CXX=${cudatoolkit.cc}/bin/g++
  '' else ''
    export CC=${gcc8Stdenv.cc}/bin/gcc CXX=${gcc8Stdenv.cc}/bin/g++
  '') + lib.optionalString (cudaSupport && cudnn != null) ''
    export CUDNN_INCLUDE_DIR=${cudnn}/include
  '';
```

It's ugly, but does the job for the time being ;).

bcdarwin commented:
Compiling NCCL as part of the Pytorch build also fails -- not sure why yet:

```
make[2]: Entering directory '/build/source/third_party/nccl/nccl/src/collectives/device'
Generating rules                               > /build/source/build/nccl/obj/collectives/device/Makefile.rules
/bin/sh: ./gen_rules.sh: not found
Generating nccl.pc.in                          > /build/source/build/nccl/lib/pkgconfig/nccl.pc
Compiling  functions.cu                        > /build/source/build/nccl/obj/collectives/device/functions.o
nvlink error   : Undefined reference to '_Z25ncclBroadcastRing_copy_i8P14CollectiveArgs' in '/build/source/build/nccl/obj/collectives/device/functions.o' (target: sm_35)
 <...tons of similar nvlink errors...>
```

Taneb commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -35,9 +35,8 @@ assert !(openMPISupport && cudaSupport) || matchesCudatoolkit openmpi;
-assert !mklSupport || mkl != null;
 assert !(mklSupport && cudaSupport) || matchesMkl magma;
-assert !mklSupport || (numpy.blasImplementation == "mkl" && numpy.blas == mkl);
+assert (mklSupport && mkl != null) || openblas != null;
```
Why have these assertions changed? I ran into a hairy situation trying to use this with a `numpy` with `mkl`.

acairncross replied:
It seems to be part of an overarching in-progress change: #74894

acairncross commented:
A couple of additional things I think this derivation should have:

- I think this should have `export MAX_JOBS=$NIX_BUILD_CORES` in somewhere like the `preConfigure` or `preBuild`, to make it respect the `--cores` `nix-build` option.
- `cp -r $out/${python.sitePackages}/torch/share   $dev/share` in the `postInstall`. This installs the CMake exported target (I think that's what it's called), and makes things like [the example CMake build here](https://pytorch.org/tutorials/advanced/cpp_frontend.html) work.

eadwu commented:
Here's the patch I use locally 
```
diff --git a/pkgs/development/python-modules/pytorch/default.nix b/pkgs/development/python-modules/pytorch/default.nix
index 6b1d72aee55..e85cd9d97b2 100644
--- a/pkgs/development/python-modules/pytorch/default.nix
+++ b/pkgs/development/python-modules/pytorch/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, fetchgit, buildPythonPackage, python, pythonOlder,
+{ stdenv, fetchurl, fetchgit, fetchpatch, buildPythonPackage, python, pythonOlder,
   cudaSupport ? false, cudatoolkit ? null, cudnn ? null, nccl ? null, magma ? null,
   mklSupport ? false, mkl ? null, openblas ? null,
   mklDnnSupport ? true,
@@ -131,6 +131,18 @@ in buildPythonPackage rec {
     export CUDNN_INCLUDE_DIR=${cudnn}/include
   '';
 
+  patches = [
+    (fetchpatch {
+      url = "https://patch-diff.githubusercontent.com/raw/pytorch/pytorch/pull/30332.patch";
+      sha256 = "1v9dwbhz3rdxcx6sz8y8j9n3bj6nqs78b1r8yg89yc15n6l4cqx2";
+    })
+
+    (fetchpatch {
+      url = "https://patch-diff.githubusercontent.com/raw/pytorch/pytorch/pull/30333.patch";
+      sha256 = "139413fl37h2fnil0cv99a67mqqnsh02k74b92by1qyr6pcfyg3q";
+    })
+  ];
+
   # Use pytorch's custom configurations
   dontUseCmakeConfigure = true;
 
@@ -219,6 +231,8 @@ in buildPythonPackage rec {
     mkdir $dev
     cp -r $out/${python.sitePackages}/torch/lib     $dev/lib
     cp -r $out/${python.sitePackages}/torch/include $dev/include
+
+    cp -r $out/${python.sitePackages}/torch/share $dev/share
   '';
 
   postFixup = stdenv.lib.optionalString stdenv.isDarwin ''
```

stites commented:
Thanks @eadwu -- I've added your patch as well as @acairncross ' suggestion that PyTorch respect nix's core count.

I noticed that python3 is failing but the delta made it seem like this was a configuration issue. How do I get access to the reports? (Python 2 should be disabled, by the way).

veprbl commented:
@GrahamcOfBorg build python37Packages.pytorch
@GrahamcOfBorg build python38Packages.pytorch

jonringer commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -122,18 +121,40 @@ in buildPythonPackage rec {
   '';
 
+  patches = [
+    (fetchpatch {
```
do you mind adding comments as to why these patches are necessary and when to remove them?
```suggestion
    # Does X, Y, and Z. Can be removed >=1.4.1
    (fetchpatch {
```

jonringer commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -122,18 +121,40 @@ in buildPythonPackage rec {
+      sha256 = "1v9dwbhz3rdxcx6sz8y8j9n3bj6nqs78b1r8yg89yc15n6l4cqx2";
+    })
+
+    (fetchpatch {
```
here as well
```suggestion
    # Does X, Y, and Z. Can be removed >=1.4.1
    (fetchpatch {
```

jonringer reviewed (commented):
otherwise LGTM

sveitser commented:
At the moment the build fails for me with 

```
Unable to checkout 6aaaa4754fe2a7467c5f0584648ba47482caad7d from https://github.com/pytorch/fbgemm.
```

More info about the cause can be found here https://github.com/pytorch/pytorch/issues/35149

yns88 commented:
pytorch has a new tag for v1.4.1 -- please check if that works for you.

stites commented:
> pytorch has a new tag for v1.4.1 -- please check if that works for you.

This branch has already scope-creeped from 1.3.0 and have had limited time to work on this. 1.4.1 should be considered an easy bump for the next person.

@s1lvester I will check on this later, fbgemm looks like an old bug. Can you please provide a derivation?

veprbl commented:
@GrahamcOfBorg build python37Packages.pytorch
@GrahamcOfBorg build python38Packages.pytorch

veprbl commented:
@stites See the failed builds. According to https://github.com/pytorch/pytorch/issues/35149#issuecomment-602803205, the v1.4.1 is supposed to fix that.

stites commented:
Merp -- it really feels like things are building on my end as I'm always ending up in a shell. I moved my compute node back to Ubuntu so perhaps it's a non-nixos problem : /

@GrahamcOfBorg build python37Packages.pytorch
@GrahamcOfBorg build python38Packages.pytorch

stites commented:
So the discrepancy between my builds and ofborg has been that my branch is reliably built, but the build breaks when this commit is cherrypicked on master. Just a heads-up for myself -- maybe someone else with a long-running PR will need to hear this, too.

I've finally gotten around to getting this fixed for nixos (reaching parity with ofborg errors). Something to be aware of is that the PRs @eadwu found did *not* get merged into pytorch-1.4.1, so the patches are still required to get this built. I've bumped the comments on those patches so that they are reexamined for v1.5.0. I've also taken myself off of the maintainer list for the time being as 2020 is looking pretty hectic for me.

One outstanding concern is that two of the previous builds failed due to a 1h timeout. I've triggered CI again, but I'm expecting two architectures to fail because of this.

stites commented:
Quick ping -- it would be nice if this could get into master before the 1.5.0 release : )

bhipple commented:
We're currently having some issues with `scipy` not building on master; see https://github.com/NixOS/nixpkgs/pull/85636 for details.  This is going to make it much easier to overlay `mkl` across the scientific computing stack, though, so we can simplify some of the `pytorch` options once it stabilizes!

bhipple commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -1,23 +1,23 @@
+{ stdenv, fetchurl, fetchgit, fetchpatch, buildPythonPackage, python, pythonOlder,
   cudaSupport ? false, cudatoolkit ? null, cudnn ? null, nccl ? null, magma ? null,
-  mklSupport ? false, mkl ? null,
+  mklSupport ? false, mkl ? null, openblas ? null,
```
We've now stabilized all of the blas and lapack provider infra in master, which means you can now rebase this and:

1. Delete the `mklSupport`, `mkl`, and `openblas` arguments
2. Add a `blas` argument, which is the virtual pkg that will be consistently instantiated with one provider across the stack.
3. When you need to figure out what it is or switch on it for doing things like adding MKL flags, you can use `blas.implementation`, which will be a string like `mkl` or `openblas`:

```
nix-repl> blas.implementation
"mkl"
```
```

bhipple commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -165,7 +192,7 @@ in buildPythonPackage rec {
   # Also of interest: pytorch ignores CXXFLAGS uses CFLAGS for both C and C++:
   # https://github.com/pytorch/pytorch/blob/v1.2.0/setup.py#L17
-  NIX_CFLAGS_COMPILE = lib.optionals (numpy.blas == mkl) [ "-Wno-error=array-bounds" ];
+  NIX_CFLAGS_COMPILE = lib.optionals mklSupport [ "-Wno-error=array-bounds" ];
```
As mentioned, drop the `mklSupport` attr from the top-level API, and replace it with a local variable:

```nix
mklSupport = blas.implementation == "mkl";
```

bhipple commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -174,33 +201,44 @@ in buildPythonPackage rec {
-  buildInputs = [
-    numpy.blas
-  ] ++ lib.optionals cudaSupport [ cudnn magma nccl ]
+  buildInputs = [(if mklSupport then mkl else openblas)]
```
```suggestion
  buildInputs = [ blas ]
```

bhipple commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -174,33 +201,44 @@ in buildPythonPackage rec {
   ] ++ lib.optionals openMPISupport [ openmpi ]
-    ++ lib.optional (pythonOlder "3.5") typing
-    ++ lib.optionals tensorboardSupport [pillow six future tensorflow-tensorboard];
+    ++ lib.optional (pythonOlder "3.5") typing;
```
```suggestion
```
Since this is disabled for python2 and the oldest python we have is python35, you can delete this.

bhipple commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -174,33 +201,44 @@ in buildPythonPackage rec {
-    ;
+  checkPhase = with lib.versions; with lib.strings; concatStringsSep " " [
+    # MKL 2019.5-only workaround. See: https://github.com/NixOS/nixpkgs/issues/75611
+    (optionalString (mklSupport && majorMinor mkl.version == "2019.5") "KMP_INIT_AT_FORK=FALSE ")
```
Master is now on `mkl-2020.1.217`; can you check if this bug is still around?  I hope we can delete this entirely.

stites replied:
I have a few healthcare-related things going on right now and am not even sure I'll be able to go through too many more cycles of review -- this will need to be reviewed by the next person.

bhipple commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -174,33 +201,44 @@ in buildPythonPackage rec {
-  checkInputs = [ hypothesis ninja ];
+  checkInputs = [ hypothesis ninja psutil ];
 
   doCheck = false; # tests take a long time for channel release, so doCheck should be overridden only when developing
```
How long are we talking?  Would be nice in general to run all the tests, particularly since this one is so complicated on variant ways to build and configure it.

stites replied:
Back when I was running build matricies in pytorch-world, it could take anywhere from 2-8 hours depending on the configuration and machine.

bhipple reviewed (changes requested):
Thanks to https://github.com/NixOS/nixpkgs/pull/83888 and the follow-up PR, we can delete a lot of the ad-hoc customization options around `mkl` and `openblas`, as they're now consistently sliced in across the whole distribution via the `blas` meta-package.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-would-you-do-if-you-were-the-bdfl-of-nix/6949/4

stites commented:
```with (import ./. {});

mkShell {
  buildInputs = [(python37.withPackages (ps: with ps; [
    (pytorch.override { blas = pkgs.blas.override {blasProvider = pkgs.mkl;}; })
  ]))];
}
```

stites commented:
@GrahamcOfBorg build python37Packages.pytorch
@GrahamcOfBorg build python38Packages.pytorch

Halfway through the build on my laptop. For mkl, I'm using this little helper script:
```nix
# in test.nix
with (import ./.  {
  config.allowUnfree = true;
  overlays = [(self: super: {
    blas = super.blas.override {
      blasProvider = super.mkl;
    };
  })];
});

mkShell {
  buildInputs = [(python37.withPackages (ps: with ps; [ pytorch ]))];
}
```
This seems correct as numpy is also building. I will point out that, before, I was attempting to override with `pytorch.override { blas = blas.override {blasProvider = mkl;}; }` and pytorch was failing to find the mkl.
 
Same as before, I'll update this comment for build updates / retries with GrahamcOfBorg and will bump the PR when it's ready for review.

stites commented:
@matthewbauer can you double-check that my script is correct and that https://github.com/NixOS/nixpkgs/pull/83888 is not messing with CMake? One small subtlety of pytorch is that it calls CMake through setup.py, but that doesn't seem like it would be an issue here.

bhipple commented:
I believe you should also set `lapack` to use `mkl` in your overlay.  Regardless of whether `pytorch` needs it, other packages in the stack do:

```nix
# in test.nix
with (import ./.  {
  config.allowUnfree = true;
  overlays = [(self: super: {
    blas = super.blas.override {
      blasProvider = super.mkl;
    };    
    lapack = super.lapack.override {
      lapackProvider = super.mkl;
    };
  })];
});

mkShell {
  buildInputs = [(python37.withPackages (ps: with ps; [ pytorch ]))];
}
```

bhipple committed:
python3Packages.pytorch: fixup builds with MKL

- Pass `blas.provider` into `buildInputs`, so that CMake can find the actual
  `mkl` for inspection of its cmake files and headers.

- Add `USE_MKL` correctly when the blas provider is `mkl`.

- Use the MKLDNN and MKLDNN_CBLAS flags by default, since `mkldnn` is FOSS and
  always available..

- Remove a patch for MKL 2019, since we've moved to 2020.

- Add a pythonImportsCheck for "torch" as a basic sanity-check

- Removed some unused variables at the top of the file

bhipple commented:
@GrahamcOfBorg build python37Packages.pytorch
 
@stites I pushed an update to your PR with the following changes:

- Pass `blas.provider` into `buildInputs`, so that CMake can find the actual
  `mkl` for inspection of its cmake files and headers.

- Add `USE_MKL` correctly when the blas provider is `mkl`.

- Use the MKLDNN and MKLDNN_CBLAS flags by default, since `mkldnn` is FOSS and
  always available..

- Remove a patch for MKL 2019, since we've moved to 2020.

- Set `doCheck` to true; this does not actually appear to take as long as it
used to. (10-15 minutes for me.)

- Removed some unused variables at the top of the file

I've managed to build this successfully with both the FOSS stack and with `mkl` as the `blas` provider; take a look and LMK what you think.

jonringer reviewed (commented):
not sure about regressions, but at the very least, we should disable for python38

```
builder for '/nix/store/a70hd44hmh22wxsr0w7sl3wnrfgfxpjb-python3.7-ignite-0.2.1.drv' failed with exit code 1; last 10 log lines:
  tests/ignite/contrib/handlers/test_param_scheduler.py::test_lr_scheduler
  tests/ignite/contrib/handlers/test_param_scheduler.py::test_lr_scheduler
  tests/ignite/contrib/handlers/test_param_scheduler.py::test_lr_scheduler
  tests/ignite/contrib/handlers/test_param_scheduler.py::test_lr_scheduler
  tests/ignite/contrib/handlers/test_param_scheduler.py::test_lr_scheduler
    /nix/store/m4i2i1ax2ch5y1ql8ng7f014mrmk63ja-python3.7-pytorch-1.4.1/lib/python3.7/site-packages/torch/optim/lr_scheduler.py:342: DeprecationWarning: To get the last learning rate computed by the scheduler, please use `get_last_lr()`.
      "please use `get_last_lr()`.", DeprecationWarning)

  -- Docs: https://docs.pytest.org/en/latest/warnings.html
  ===== 4 failed, 275 passed, 4 skipped, 72 deselected, 20 warnings in 6.45s =====
builder for '/nix/store/zyxn5nxx22jwdj5xhvxx173ngkms6ryi-python3.8-pytorch-1.4.1.drv' failed with exit code 1; last 10 log lines:
  ----------------------------------------------------------------------
  Ran 2276 tests in 93.630s

  FAILED (failures=1, skipped=66, expected failures=1)
  Traceback (most recent call last):
    File "test/run_test.py", line 455, in <module>
      main()
    File "test/run_test.py", line 448, in main
      raise RuntimeError(message)
  RuntimeError: test_jit failed!
cannot build derivation '/nix/store/iraap11hf3k4yax6bzpmnmrif3pzli5l-python3.8-ignite-0.2.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/8qdsp4ga63vmwbq8w0b1vf33avb848rx-python3.8-tensorly-0.4.5.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/qkmv9p848yh1l62cn9yy3wb6id0hh1ac-python3.8-torchvision-0.2.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/11waqaqv5xq4276f1npv1dmpj31hgprx-python3.8-pywick-0.5.6.drv': 2 dependencies couldn't be built
builder for '/nix/store/c8jvyrqblrk4a95rlnv5ji142d7sv33m-python3.7-pytorch-1.4.1.drv' failed with exit code 1; last 10 log lines:
  wrapping `/nix/store/28cbbgvg7mm81l4q7qac7pmmqycjizj2-python3.7-pytorch-1.4.1/bin/convert-caffe2-to-onnx'...
  Executing pythonRemoveTestsDir
  Finished executing pythonRemoveTestsDir
  running install tests
  Traceback (most recent call last):
    File "test/run_test.py", line 14, in <module>
      import torch
    File "/nix/store/28cbbgvg7mm81l4q7qac7pmmqycjizj2-python3.7-pytorch-1.4.1/lib/python3.7/site-packages/torch/__init__.py", line 81, in <module>
      from torch._C import *
  ImportError: /nix/store/28cbbgvg7mm81l4q7qac7pmmqycjizj2-python3.7-pytorch-1.4.1/lib/python3.7/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch4cuda4nccl6detail16throw_nccl_errorE12ncclResult_t
builder for '/nix/store/f2lxq1zcw1zbc4q3snnskdaadm5vyw2x-python3.8-pytorch-1.4.1.drv' failed with exit code 1; last 10 log lines:
  wrapping `/nix/store/q9diyfdsl34cy1iqcnkg7c6g650aif52-python3.8-pytorch-1.4.1/bin/convert-caffe2-to-onnx'...
  Executing pythonRemoveTestsDir
  Finished executing pythonRemoveTestsDir
  running install tests
  Traceback (most recent call last):
    File "test/run_test.py", line 14, in <module>
      import torch
    File "/nix/store/q9diyfdsl34cy1iqcnkg7c6g650aif52-python3.8-pytorch-1.4.1/lib/python3.8/site-packages/torch/__init__.py", line 81, in <module>
      from torch._C import *
  ImportError: /nix/store/q9diyfdsl34cy1iqcnkg7c6g650aif52-python3.8-pytorch-1.4.1/lib/python3.8/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch4cuda4nccl6detail16throw_nccl_errorE12ncclResult_t
cannot build derivation '/nix/store/rqcqydv29f29gn957kfb5vysx4ppgnsv-env.drv': 8 dependencies couldn't be built
[0 built (4 failed), 0.0 MiB DL]
error: build of '/nix/store/rqcqydv29f29gn957kfb5vysx4ppgnsv-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/75827
2 package marked as broken and skipped:
python37Packages.pyro-ppl python38Packages.pyro-ppl

8 package failed to build:
python37Packages.ignite python37Packages.pytorchWithCuda python38Packages.ignite python38Packages.pytorch python38Packages.pytorchWithCuda python38Packages.pywick python38Packages.tensorly python38Packages.torchvision

4 package built:
python37Packages.pytorch python37Packages.pywick python37Packages.tensorly python37Packages.torchvision
```

bhipple commented:
Hmm, I think I'll go back to just not running the test suite -- particularly since users of MKL may be running without a binary cache and need to wait for the package to build themselves.

jonringer commented:
or just reduce the checkphase to something simple. I would be fine with some `pythonImportsCheck = [ "torch" ];`.

As long as the maintainer verified that it worked for more in-depth cases locally

stites commented:
:+1: thanks for bumping this pr @bhipple !

bhipple committed:
python3Packages.ignite: 0.2.1 -> 0.3.0

bhipple commented:
Result of `nixpkgs-review pr 75827` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
<br>- python37Packages.pyro-ppl
<br>- python38Packages.pyro-ppl
</details>
<details>
  <summary>4 packages failed to build:</summary>
<br>- python37Packages.ignite
<br>- python37Packages.pytorchWithCuda
<br>- python38Packages.ignite
<br>- python38Packages.pytorchWithCuda
</details>
<details>
  <summary>8 packages built:</summary>
<br>- python37Packages.pytorch (python37Packages.pytorchWithoutCuda)
<br>- python37Packages.pywick
<br>- python37Packages.tensorly
<br>- python37Packages.torchvision
<br>- python38Packages.pytorch (python38Packages.pytorchWithoutCuda)
<br>- python38Packages.pywick
<br>- python38Packages.tensorly
<br>- python38Packages.torchvision
</details>

bhipple commented:
`pytorchWithCuda` is broken on `master`; help requested, but since it's not a regression relative to this PR, and since this PR has significant improvements to `pytorch`, I propose merging as-is and fixing cuda in another PR.

`ignite` required an upgrade to keep it building.

bhipple committed:
python3Packages.pytorch: reference dnnl by new name oneDNN

bhipple commented:
One last change: in https://github.com/NixOS/nixpkgs/pull/85839 I changed the name of `dnnl` to `oneDNN` to track an upstream rename.  That PR has a back-compat alias as `dnnl`, but it looks like OfBorg builds its eval without aliases (which is probably a good thing in general).

Updated pytorch to reference `dnnl` by the new name, but made no other material change.

bhipple commented:
Result of `nixpkgs-review pr 75827` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
<br>- python37Packages.pytorchWithCuda
<br>- python38Packages.pytorchWithCuda
</details>
<details>
  <summary>10 packages built:</summary>
<br>- python37Packages.ignite
<br>- python37Packages.pytorch (python37Packages.pytorchWithoutCuda)
<br>- python37Packages.pywick
<br>- python37Packages.tensorly
<br>- python37Packages.torchvision
<br>- python38Packages.ignite
<br>- python38Packages.pytorch (python38Packages.pytorchWithoutCuda)
<br>- python38Packages.pywick
<br>- python38Packages.tensorly
<br>- python38Packages.torchvision
</details>

Zhen-hao commented:
great work! thank you all!
I've been waiting for this

Zhen-hao commented:
what is the best way to get `pytorchWithCuda` working on NixOS? 
currently, the build fails with 
```
    from torch._C import *
ImportError: /nix/store/a29i85licwqsbvysansiyd67zwlkianb-python3.7-pytorch-1.4.1/lib/python3.7/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch4cuda4nccl6detail16throw_nccl_errorE12ncclResult_t
```
for now I will try to follow https://github.com/NixOS/nixpkgs/pull/75827#issuecomment-582350469

jonringer commented:
that looks like a linking error to cuda. Not super familiar with the cuda toolchain to know where an assumption is being broken

bhipple commented:
If you manage to get it working, please do send a PR!

bhipple commented:
Interesting note: when we upgraded pytorch from 1.0.0 -> 1.2.0, we must've leaked a proprietary dependency into the default expression, because Hydra stopped building it.  That's now been fixed with this PR, so we once again have binary cache builds of the default pytorch:

https://hydra.nixos.org/job/nixpkgs/trunk/python38Packages.pytorch.x86_64-linux/all

Zhen-hao commented:
> what is the best way to get `pytorchWithCuda` working on NixOS?
> currently, the build fails with
> 
> ```
>     from torch._C import *
> ImportError: /nix/store/a29i85licwqsbvysansiyd67zwlkianb-python3.7-pytorch-1.4.1/lib/python3.7/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch4cuda4nccl6detail16throw_nccl_errorE12ncclResult_t
> ```
> 
> for now I will try to follow [#75827 (comment)](https://github.com/NixOS/nixpkgs/pull/75827#issuecomment-582350469)

I tried a few things but couldn't make it work with the current version. 
However, it builds with CUDA when I change the version to 1.5.0 and remove the `patches` list. I tested it with the examples in transformers' quickstart guide and it worked fine with my GPU. My build uses `cudnn_cudatoolkit_10_2`.
I think it was bug in pytorch 1.4.1 and they fixed in 1.5.0.
